File: Striver Important Interview Questions/Day 2 - Arrays Part 2/find-the-duplicate-number.cpp

// https://leetcode.com/problems/find-the-duplicate-number/description/

class Solution {
public:
    int findDuplicate(vector<int>& nums) {
        
        int slow = nums[0];
        slow = nums[slow];
        int fast = nums[0];
        fast = nums[nums[fast]];

        while(slow != fast){
            slow = nums[slow];
            fast = nums[nums[fast]];
        }

        fast = nums[0];
        while(slow != fast){
            slow = nums[slow];
            fast = nums[fast];
        }
        return fast;
    }
};

File: Neetcode-150-js/Linked List/find-the-duplicate-number.js

// https://leetcode.com/problems/find-the-duplicate-number/description/

/**
 * @param {number[]} nums
 * @return {number}
 */
var findDuplicate = function(nums) {
    let slow = nums[0];
    slow = nums[slow];

    let fast = nums[0];
    fast = nums[nums[fast]];

    while(slow !== fast){
        slow = nums[slow];
        fast = nums[nums[fast]];
    }

    fast = nums[0];
    while(slow !== fast){
        slow = nums[slow];
        fast = nums[fast];
    }

    return slow;
};